pkg/plugins/localfiles.go

```diff
@@ -10,79 +11,189 @@ import (
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+		acc[path] = struct{}{}
+		return nil
+	}
+}
+
+// emptyCollectFilesFunc is a no-op CollectFilesFunc. It returns a filepath.WalkFunc that does nothing and
+// the accumulator map is not altered in any way. This is suitable to use with plugins.allowListFSNoFiles for tests.
+func emptyCollectFilesFunc(_ map[string]struct{}) filepath.WalkFunc {
+	return func(path string, info fs.FileInfo, err error) error {
+		return nil
+	}
+}
 
 // LocalFS is a plugins.FS that allows accessing files on the local file system.
 type LocalFS struct {
-	// m is a map of relative file paths that can be accessed on the local filesystem.
-	// The path separator must be os-specific.
-	m map[string]*LocalFile
-
-	// basePath is the basePath that will be prepended to all the files (in m map) before accessing them.
+	// basePath is the basePath that will be prepended to all the files to get their absolute path.
 	basePath string
-}
 
-// NewLocalFS returns a new LocalFS that can access the specified files in the specified base path.
-// Both the map keys and basePath should use the os-specific path separator for Open() to work properly.
-func NewLocalFS(m map[string]struct{}, basePath string) LocalFS {
-	pfs := make(map[string]*LocalFile, len(m))
-	for k := range m {
-		pfs[k] = &LocalFile{
-			path: k,
-		}
-	}
+	// collectFilesFunc returns a filepath.WalkFunc that creates a slice of files in this LocalFS, for Files()
+	collectFilesFunc CollectFilesFunc
+}
 
-	return LocalFS{
-		m:        pfs,
-		basePath: basePath,
+// NewLocalFS returns a new LocalFS that can access any file in the specified base path on the filesystem.
+// basePath must use os-specific path separator for Open() to work properly.
+// If collectFilesFunc is not provided, defaultCollectFilesFunc is used instead, which will collect all files.
+func NewLocalFS(basePath string, collectFilesFunc CollectFilesFunc) LocalFS {
+	if collectFilesFunc == nil {
+		collectFilesFunc = defaultCollectFilesFunc
 	}
+	return LocalFS{basePath: basePath, collectFilesFunc: collectFilesFunc}
 }
 
-// Open opens the specified file on the local filesystem, and returns the corresponding fs.File.
-// If a nil error is returned, the caller should take care of closing the returned file.
+// Open opens the specified file on the local filesystem.
+// The provided name must be a relative file name that uses os-specific path separators.
+// The function returns the corresponding fs.File for accessing the file on the local filesystem.
+// If a nil error is returned, the caller should take care of calling Close() the returned fs.File.
+// If the file does not exist, ErrFileNotExist is returned.
 func (f LocalFS) Open(name string) (fs.File, error) {
 	cleanPath, err := util.CleanRelativePath(name)
-	if err != nil {
-		return nil, err
+	if strings.Contains(cleanPath, "..") || err != nil {
+		return nil, ErrFileNotExist
```

Comment:
should this error be something different than "not exists"?

Reply:
Ops, good point, it should have been:

```go
if err != nil {
	return nil, err
}
if strings.Contains(cleanPath, "..") {
	return nil, ErrFileNotExist
}
```

I'll change that :grinning:

Reply:
well, true, but I meant that having '..' in the path mean that the path is invalid not that it doesn't exist, no?

Reply:
God rid of this error case in the latest version as it was pretty much already handled by `CleanRelativePath`